text/0000-route-prefix.md

```diff
@@ -0,0 +1,109 @@
+
+Providing the capability to configure a route prefix facilitates decoupling these concerns, allowing a web app to more easily be deployed without concern for the host path.  As an example, configuration of a route prefix would allow for the registering '/safety' the work seamlessly for both 'https://drive.uber.com/safety' and 'https://uber.com/drive/safety'.
+
+# Detailed design
+
+The impementation involves three main concerns:
+
+* Static asset pathing
+* `ctx.url` re-write
+* Prefix handling for `fusion-plugin-reactroute`
+
+### Static asset pathing
+
+`fusion-cli` currently handles supplying the static asset path via the environment variables plugin.  Values are supplied via [process.env](https://nodejs.org/api/process.html#process_process_env) with reasonable fallback defaults.
+
+For example, the static asset pathing is defined by the key, `FRAMEWORK_STATIC_ASSET_PATH`, with a default value of `/_static`.
+
+We can provide the route prefix in a similar fashion, and update static asset path resolution accordingly:
+
+* Passing `ROUTE_PREFIX` as an environment variable that can be accessed via the environment variables plugin.
+* Ensure that the default static asset path incorporates the route prefix (e.g. '/privacy/_static', where '/privacy' is the route prefix).
+
+In production, this will likely be supplied via an app's `pinocchio.yaml` configuration file:
+
+```
+website:
+  command: 'ROUTE_PREFIX="/privacy" FRAMEWORK_STATIC_ASSET_PATH="https://someuuid.cloudfront.net/uberprivacy" npm start'
+```
+
+Ideally, as this is being provided as an environment variable, we should scope this to the `fusion-cli` and ensure it is not exposed to user land.
+
+Note that the primary [webpack considerations](https://github.com/fusionjs/fusion-cli/search?utf8=%E2%9C%93&q=__webpack_public_path__&type=) here are already handled via the [`compilation-metadata-plugin`](https://github.com/fusionjs/fusion-cli/blob/master/plugins/compilation-metadata-plugin.js) plugin.
+
+### `ctx.url` rewrite
+
+Many middleware are expected to run conditional checks against `ctx.url`.  It is not the concern of the middleware to know whether the web app is deployed at the root of a host, or behind a route prefix.  In order to enforce this, a core middleware can strip the route prefix from `ctx.url` prior to any of the user-land middlewares' execution.  This has the benefit of allowing middlewares to be easily re-used across applications regardless of deployment concerns.
+
+e.g.
+```
+# fusion-cli: src/route-prefix-plugin.js
+const envVarsPlugin = require('./environment-variables-plugin');
+module.exports = function() {
+  const envVars = envVarsPlugin().of();
+  return function middleware(ctx, next) {
+    const prefix = envVars.prefix;
+
+    // enhance ctx.url sans prefix
+    if(ctx.url.indexOf(prefix) === 0 /*found at index 0*/) {
+      ctx.url = ctx.url.slice(prefix.length);
+    }
+  }
+}
+```
+
+### Prefix handling for `fusion-plugin-react-router`
```

Comment:
@mlmorg noted that there may exist other user-land plugins that will require the route prefix as a configuration property.  We'll want to identify all of these are part of this process.